File: dotnet/src/Util/MongoCrypt.cs
using MongoDB.Bson;
using MongoDB.Driver;
using MongoDB.Driver.Encryption;
using System;
using System.Collections.Generic;
using System.Threading;
using System.Threading.Tasks;

namespace MongoEncryption.Util
{
    internal class MongoCrypt
    {
        private CollectionNamespace _vaultNamespace = null;
        private Dictionary<string, IReadOnlyDictionary<string, object>> _kmsProviders = null;
        private ClientEncryption _clientEncryption = null;
        private string _vaultEndpoint = String.Empty;
        private string _keyName = String.Empty;

        private string _clientConnectionString = String.Empty;
        private MongoClient _autoDecryptionClient = null;
        private MongoClient _defaultClient = null;

        public MongoCrypt(string connectionString, string keyName)
        {
            _clientConnectionString = connectionString;
            _keyName = keyName;

            var vaultConnectionString = Environment.GetEnvironmentVariable("mongoVaultConnection");
            _vaultNamespace = CollectionNamespace.FromFullName(Environment.GetEnvironmentVariable("mongoVaultNamespace"));
            _vaultEndpoint = Environment.GetEnvironmentVariable("vaultEndpoint");

            _kmsProviders = new Dictionary<string, IReadOnlyDictionary<string, object>>();

            //Get service principal details from config
            var azureTenantId = Environment.GetEnvironmentVariable("encryptionPrincipalTenantId");
            var azureClientId = Environment.GetEnvironmentVariable("encryptionPrincipalClientId");
            var azureClientSecret = Environment.GetEnvironmentVariable("encryptionPrincipalClientSecret");

            //Build Dictionary options required to access Key Vault from the driver
            var azureKmsOptions = new Dictionary<string, object>
            {
                { "tenantId", azureTenantId },
                { "clientId", azureClientId },
                { "clientSecret", azureClientSecret },
            };

            //Map Azure provider and options
            _kmsProviders.Add("azure", azureKmsOptions);

            //Build Mongo connection string for encryption collection
            var vaultClient = new MongoClient(vaultConnectionString);

            //Build client encryption options
            var clientEncryptionOptions = new ClientEncryptionOptions(
                keyVaultClient: vaultClient,
                keyVaultNamespace: _vaultNamespace,
                kmsProviders: _kmsProviders);
            
            //Create client encryption instance
            _clientEncryption = new ClientEncryption(clientEncryptionOptions);
        }

        public MongoClient AutoDecryptionClient
        {
            get
            {
                if (_autoDecryptionClient == null)
                {
                    //Build auto encryption settings, providing vault details
                    var clientSettingsAutoDecrypt = MongoClientSettings.FromConnectionString(_clientConnectionString);
                    var autoEncryptionOptions = new AutoEncryptionOptions(
                        keyVaultNamespace: _vaultNamespace,
                        kmsProviders: _kmsProviders,
                        bypassAutoEncryption: true); //Must be true to skip auto encryption (not supported on Mongo Community or Cosmos DB)
                    clientSettingsAutoDecrypt.AutoEncryptionOptions = autoEncryptionOptions;

                    //Create mongo client instance with encryption settings
                    _autoDecryptionClient = new MongoClient(clientSettingsAutoDecrypt);
                }

                return _autoDecryptionClient;
            }
        }

        public MongoClient DefaultClient
        {
            get
            {
                //Create mongo client instance without encryption settings
                if (_defaultClient == null)
                    _defaultClient = new MongoClient(_clientConnectionString);

                return _defaultClient;
            }
        }

        public async Task<Guid> CreateKeyAsync()
        {
            //Calls Key Vault key creation
            var keyVersion = await KeyVaultOperations.CreateKeyAsync(_vaultEndpoint, _keyName);

            //Create data key options with key details to store in mongo collection
            var dataKeyOptions = new DataKeyOptions(
                            masterKey: new BsonDocument
                            {
                                { "keyName", _keyName },
                                { "keyVaultEndpoint", _vaultEndpoint },
                                { "keyVersion", keyVersion }
                            },
                            alternateKeyNames: new List<string>()
                            {
                                { _keyName } //AltKey: friendly name to retrieve the key
                            }
                            );

            //Stores key details in mongo collection
            var dataKeyId = await _clientEncryption.CreateDataKeyAsync("azure", dataKeyOptions);

            return dataKeyId;
        }


        public async Task<bool> RotateKeyAsync()
        {
            //Calls Key Vault key rotation
            var keyVersion = await KeyVaultOperations.RotateKeyAsync(_vaultEndpoint, _keyName);

            //Create rewrap data key options with key details to store in mongo collection (new key version)
            var rewrapDataKey = new RewrapManyDataKeyOptions("azure",
                masterKey: new BsonDocument
                {
                    { "keyName", _keyName },
                    { "keyVaultEndpoint", _vaultEndpoint },
                    { "keyVersion", keyVersion }
                }
                );

            //Rewraps DEK and stores new version in mongo collection
            var dataKey = await _clientEncryption.RewrapManyDataKeyAsync(new BsonDocument { { "masterKey.keyName", _keyName } }, rewrapDataKey);

            return dataKey.BulkWriteResult.IsAcknowledged;
        }

        public async Task<BsonValue> EncryptFieldAsync(BsonValue field, bool isDeterministic = false)
        {
            //Explicit Client Encryption used for storing encrypted data or encrypting deterministic search fields
            var encryptedField = await _clientEncryption.EncryptAsync(
                field,
                new EncryptOptions(algorithm: (isDeterministic ? "AEAD_AES_256_CBC_HMAC_SHA_512-Deterministic" : "AEAD_AES_256_CBC_HMAC_SHA_512-Random"),
                    alternateKeyName: _keyName),
                CancellationToken.None);

            return encryptedField;
        }
    }
}

File: java/src/main/java/com/function/util/MongoCrypt.java
package com.function.util;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.Map;

import org.bson.BsonBinary;
import org.bson.BsonDocument;
import org.bson.BsonString;
import org.bson.BsonValue;

import com.mongodb.AutoEncryptionSettings;
import com.mongodb.ClientEncryptionSettings;
import com.mongodb.ConnectionString;
import com.mongodb.MongoClientSettings;
import com.mongodb.client.MongoClient;
import com.mongodb.client.MongoClients;
import com.mongodb.client.model.Filters;
import com.mongodb.client.model.vault.DataKeyOptions;
import com.mongodb.client.model.vault.EncryptOptions;
import com.mongodb.client.model.vault.RewrapManyDataKeyOptions;
import com.mongodb.client.vault.ClientEncryption;
import com.mongodb.client.vault.ClientEncryptions;

public class MongoCrypt {

    private String _vaultNamespace = null;
    private Map<String, Map<String, Object>> _kmsProviders = null;
    private ClientEncryption _clientEncryption = null;
    private String _vaultEndpoint = null;
    private String _keyName = null;

    private String _clientConnectionString = null;
    private MongoClient _autoDecryptionClient = null;
    private MongoClient _defaultClient = null;

    public MongoCrypt(String connectionString, String keyName)
    {
        _clientConnectionString = connectionString;
        _keyName = keyName;

        var vaultConnectionString = System.getenv("mongoVaultConnection");
        _vaultNamespace = System.getenv("mongoVaultNamespace"); //database.collection used to store key references
        _vaultEndpoint = System.getenv("vaultEndpoint");

        _kmsProviders = new HashMap<String, Map<String, Object>>();

        //Get service principal details from config
        var azureTenantId = System.getenv("encryptionPrincipalTenantId");
        var azureClientId = System.getenv("encryptionPrincipalClientId");
        var azureClientSecret = System.getenv("encryptionPrincipalClientSecret");

        //Build hashmap options required to access Key Vault from the driver
        var azureKmsOptions = new HashMap<String, Object>() {{
            put("tenantId", azureTenantId);
            put("clientId", azureClientId);
            put("clientSecret", azureClientSecret);
        }};

        //Map Azure provider and options
        _kmsProviders.put( "azure", azureKmsOptions);

        //Build Mongo connection string for encryption collection
        var kvmcs = MongoClientSettings.builder()
                                        .applyConnectionString(new ConnectionString(vaultConnectionString))
                                        .build();

        //Build client encryption options
        var clientEncryptionOptions = ClientEncryptionSettings.builder()
                                                               .keyVaultMongoClientSettings(kvmcs)
                                                               .keyVaultNamespace(_vaultNamespace)
                                                               .kmsProviders(_kmsProviders)
                                                               .build();

        //Create client encryption instance
        _clientEncryption = ClientEncryptions.create(clientEncryptionOptions);
    }

    public MongoClient AutoDecryptionClient()
    {
        if (_autoDecryptionClient == null)
        {
            //Build auto encryption settings, providing vault details
            var autoEncryptionOptions = AutoEncryptionSettings.builder()
                                            .keyVaultNamespace(_vaultNamespace)
                                            .kmsProviders(_kmsProviders)
                                            .bypassAutoEncryption(true) //Must be true to skip auto encryption (not supported on Mongo Community or Cosmos DB)
                                            .build();

            //Build mongo client settings
            var kvmcs = MongoClientSettings.builder()
                                            .applyConnectionString(new ConnectionString(_clientConnectionString))
                                            .autoEncryptionSettings(autoEncryptionOptions)
                                            .build();

            //Create mongo client instance with encryption settings
            _autoDecryptionClient = MongoClients.create(kvmcs);
        }

        return _autoDecryptionClient;
    }

    public MongoClient DefaultClient()
    {
        //Create mongo client instance without encryption settings
        if (_defaultClient == null)
            _defaultClient = MongoClients.create(MongoClientSettings.builder()
                                                                        .applyConnectionString(new ConnectionString(_clientConnectionString))
                                                                        .build());
        
        return _defaultClient;
    }

    public String CreateKey()
    {
        //Calls Key Vault key creation
        var keyVersion = KeyVaultOperations.CreateKey(_vaultEndpoint, _keyName);

        //Build BSON document with key details to store in mongo collection
        var masterKey = new BsonDocument();
        masterKey.append("keyName", new BsonString(_keyName))
                .append("keyVaultEndpoint", new BsonString(_vaultEndpoint))
                .append("keyVersion", new BsonString(keyVersion));

        var altKey = new ArrayList<String>();
        altKey.add(_keyName); //AltKey: friendly name to retrieve the key

        //Create data key options
        var dataKeyOptions = new DataKeyOptions().masterKey(masterKey).keyAltNames(altKey);

        //Stores key details in mongo collection
        var dataKeyId = _clientEncryption.createDataKey("azure", dataKeyOptions);

        return dataKeyId.asUuid().toString();
    }

    public Boolean RotateKey()
    {
        //Calls Key Vault key rotation
        var keyVersion = KeyVaultOperations.RotateKey(_vaultEndpoint, _keyName);

        //Build BSON document with key details to store in mongo collection (new key version)
        var rewrapDataKey = new BsonDocument();
        rewrapDataKey.append("keyName", new BsonString(_keyName))
                .append("keyVaultEndpoint", new BsonString(_vaultEndpoint))
                .append("keyVersion", new BsonString(keyVersion));

        //Create rewrap data key options
        var rewrapDataKeyOptions = new RewrapManyDataKeyOptions().provider("azure").masterKey(rewrapDataKey);

        //Rewraps DEK and stores new version in mongo collection
        var dataKey = _clientEncryption.rewrapManyDataKey(Filters.eq("masterKey.keyName", _keyName), rewrapDataKeyOptions);

        return dataKey.getBulkWriteResult().getModifiedCount() != 0;
    }

    public BsonBinary EncryptField(BsonValue field, Boolean isDeterministic)
    {
        //Explicit Client Encryption used for storing encrypted data or encrypting deterministic search fields
        var encryptedField = _clientEncryption.encrypt(
            field,
            new EncryptOptions((isDeterministic ? "AEAD_AES_256_CBC_HMAC_SHA_512-Deterministic" : "AEAD_AES_256_CBC_HMAC_SHA_512-Random")).keyAltName(_keyName));

        return encryptedField;
    }
}
